pkg/observability/k8s-events.go

```diff
@@ -0,0 +1,178 @@
+	PostDrainMsg            = "Post-drain task successfully executed"
+)
+
+// Interruption event reasons
+const (
+	scheduledEventReason          = "ScheduledEvent"
+	spotITNReason                 = "SpotInterruption"
+	sqsTerminateReason            = "SQSTermination"
+	rebalanceRecommendationReason = "RebalanceRecommendation"
+	unknownReason                 = "UnknownInterruption"
+)
+
+// K8sEventRecorder wraps a Kubernetes event recorder with some extra information
+type K8sEventRecorder struct {
+	annotations map[string]string
+	enabled     bool
+	node        *corev1.Node
+	record.EventRecorder
+}
+
+// InitK8sEventRecorder creates a Kubernetes event recorder
+func InitK8sEventRecorder(enabled bool, nodeName string, nodeMetadata ec2metadata.NodeMetadata, extraAnnotationsStr string) (K8sEventRecorder, error) {
+	if !enabled {
+		return K8sEventRecorder{}, nil
+	}
+
+	// Create default annotations
+	// Worth iterating over nodeMetadata fields using reflect? (trutx)
+	annotations := make(map[string]string)
+	annotations["account-id"] = nodeMetadata.AccountId
+	annotations["availability-zone"] = nodeMetadata.AvailabilityZone
+	annotations["instance-id"] = nodeMetadata.InstanceID
+	annotations["instance-type"] = nodeMetadata.InstanceType
+	annotations["local-hostname"] = nodeMetadata.LocalHostname
+	annotations["local-ipv4"] = nodeMetadata.LocalIP
+	annotations["public-hostname"] = nodeMetadata.PublicHostname
+	annotations["public-ipv4"] = nodeMetadata.PublicIP
+	annotations["region"] = nodeMetadata.Region
+
+	// Parse extra annotations
+	var err error
+	if extraAnnotationsStr != "" {
+		annotations, err = parseExtraAnnotations(annotations, extraAnnotationsStr)
+		if err != nil {
+			return K8sEventRecorder{}, err
+		}
+	}
+
+	// Get in-cluster config
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		return K8sEventRecorder{}, err
+	}
+
+	// Create clientSet
+	clientSet, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return K8sEventRecorder{}, err
+	}
+
+	// Get node
+	node, err := clientSet.CoreV1().Nodes().Get(nodeName, metav1.GetOptions{})
+	if err != nil {
+		return K8sEventRecorder{}, err
+	}
+
+	// Create broadcaster
+	broadcaster := record.NewBroadcaster()
+	broadcaster.StartRecordingToSink(&typedcorev1.EventSinkImpl{Interface: clientSet.CoreV1().Events("default")})
+
+	// Create event recorder
+	return K8sEventRecorder{
+		annotations: annotations,
+		enabled:     true,
+		node:        node,
+		EventRecorder: broadcaster.NewRecorder(
+			scheme.Scheme,
+			corev1.EventSource{
+				Component: "aws-node-termination-handler",
+				Host:      nodeName,
+			},
+		),
+	}, nil
+}
+
+// Emit a Kubernetes event for the current node and with the given event type, reason and message
+func (r K8sEventRecorder) Emit(eventType, eventReason, eventMsgFmt string, eventMsgArgs ...interface{}) {
+	if r.enabled {
+		r.AnnotatedEventf(r.node, r.annotations, eventType, eventReason, eventMsgFmt, eventMsgArgs...)
```

Comment:
in queue processor mode, `node` might not be the node that's being cordoned and drained, is that intentional? perhaps we should pass in `node` ?

Reply:
Good point, totally omitted that. Revamped it so instead of getting the Node from the API when initialising the recorder we just create an `ObjectReference` to a `Node` object with a parameterised name. Such name will be passed in every time `recorder.Emit()` is called and it will come from the `NodeName` stored in every particular interruption event, thus making it valid both for IMDS and SQS processors if I'm not mistaken.